crates/vm/levm/src/vm.rs

```diff
@@ -379,10 +379,37 @@ impl VM {
         }
     }
 
+    /// Based on ethereum yellowpaper initial tests of intrinsic validity (Section 6), which last version is
+    /// Shanghai, so there are probably missing cancun validations.
+    fn validate_transaction(&self) -> Result<(), VMError> {
+        // Validations (1), (2), (3) and (5) are done in upper layers.
+
+        let sender_account = match self.db.accounts.get(&self.env.origin) {
+            Some(acc) => acc,
+            None => return Err(VMError::SenderAccountDoesNotExist),
+        };
+
+        // (4) the sender account has no contract code deployed (see EIP-3607 by Feist et al. [2021]);
+        if sender_account.has_code() {
+            return Err(VMError::SenderAccountShouldNotHaveBytecode);
+        }
+
+        // (6) the sender account balance contains at least the cost, v0, required in up-front payment;
+        if sender_account.balance < self.call_frames[0].msg_value {
+            return Err(VMError::SenderBalanceShouldContainTransferValue);
+        }
+
+        //(7) maxFeePerGas (t2 transactions) or gasPrice (t0 or y1 transactions), is greater than or equal to the block’s base fee
+        if self.env.gas_price < self.env.base_fee_per_gas {
+            return Err(VMError::GasPriceIsLowerThanBaseFee);
+        }
+
+        Ok(())
+    }
```

Comment:
I think that would be easier to do that validation in the other side rather than in the VM side because I had trouble determining if a transaction was type 2 or not. A way of validating it that would be:
``` Rust
// (8) for type 2 transactions, maxPriorityFeePerGas, Tf , must be no larger than maxFeePerGas, Tm.
if is_type_2_transaction() && (self.env.max_priority_fee_per_gas > self.env.gas_price) {
        return Err(VMError::MaxPriorityFeePerGasIsLargerThanMaxFeePerGas);
}
```

Reply:
Good point. Let's explain this as a comment.